src/graph/sampler.cc

```diff
@@ -988,40 +1007,59 @@ NegSubgraph NegEdgeSubgraph(GraphPtr gptr, IdArray relations, const Subgraph &po
   dgl_id_t *neg_eid_data = static_cast<dgl_id_t *>(neg_eid->data);
   dgl_id_t *induced_neg_eid_data = static_cast<dgl_id_t *>(induced_neg_eid->data);
 
+  const dgl_id_t *unchanged;
+  dgl_id_t *neg_unchanged;
+  dgl_id_t *neg_changed;
+  if (is_neg_head_mode(neg_mode)) {
+    unchanged = dst_data;
+    neg_unchanged = neg_dst_data;
+    neg_changed = neg_src_data;
+  } else {
+    unchanged = src_data;
+    neg_unchanged = neg_src_data;
+    neg_changed = neg_dst_data;
+  }
+
   dgl_id_t curr_eid = 0;
   std::vector<size_t> neg_vids;
   neg_vids.reserve(neg_sample_size);
   std::unordered_map<dgl_id_t, dgl_id_t> neg_map;
+  // If we don't exclude positive edges, we are actually sampling more than
+  // the total number of nodes in the graph.
+  if (!exclude_positive && neg_sample_size >= num_tot_nodes) {
+    // We add all nodes as negative nodes.
+    for (int64_t i = 0; i < num_tot_nodes; i++) {
+      neg_vids.push_back(i);
+      neg_map[i] = i;
+    }
+  }
+
   for (int64_t i = 0; i < num_pos_edges; i++) {
     size_t neg_idx = i * neg_sample_size;
-    neg_vids.clear();
 
     std::vector<size_t> neighbors;
     DGLIdIters neigh_it;
-    const dgl_id_t *unchanged;
-    dgl_id_t *neg_unchanged;
-    dgl_id_t *neg_changed;
     if (is_neg_head_mode(neg_mode)) {
-      unchanged = dst_data;
-      neg_unchanged = neg_dst_data;
-      neg_changed = neg_src_data;
       neigh_it = gptr->PredVec(induced_vid_data[unchanged[i]]);
     } else {
-      unchanged = src_data;
-      neg_unchanged = neg_src_data;
-      neg_changed = neg_dst_data;
       neigh_it = gptr->SuccVec(induced_vid_data[unchanged[i]]);
     }
 
-    if (exclude_positive) {
+    // If the number of negative nodes is smaller than the number of total nodes
+    // in the graph.
+    if (exclude_positive && neg_sample_size < num_tot_nodes) {
       std::vector<size_t> exclude;
       for (auto it = neigh_it.begin(); it != neigh_it.end(); it++) {
         dgl_id_t global_vid = *it;
         exclude.push_back(global_vid);
       }
+      neg_vids.clear();
       RandomSample(num_tot_nodes, neg_sample_size, exclude, &neg_vids);
-    } else {
+    } else if (neg_sample_size < num_tot_nodes) {
+      neg_vids.clear();
       RandomSample(num_tot_nodes, neg_sample_size, &neg_vids);
+    } else if (exclude_positive) {
```

Comment:
If it possible that all the above conditions are false?

Reply:
I see it's impossible. Then just use `else`

Reply:
it is possible. Sometimes, a user wants to sample negative edges of the same number as the number of nodes in a graph. at the same time, they want to turn on `exclude_positive`.

Reply:
Then what happens if both `exclude_positive` and `neg_sample_size<num_tot_nodes` are false? Do nothing? If that's the case, I would suggest have an extra else block to make it clearer.
```c++
} else {
  // do nothing (with little explanation)
}
```

Reply:
Oh, another question. You added `neg_vids.clear()` in some branches. Is it correct? The next statement is `RandomSample(..., &neg_vids)`, which I assume it adds some vids to `neg_vids`. Wouldn't it be cleared in the next iteration?

Reply:
Oh, never mind. Saw the `neg_vids.clear()` in the original code. It looks like you want to reuse the buffer. Take back my last comment.

Reply:
OK. I'll add an else block and add comment.